libraries/AP_BoardConfig/AP_BoardConfig_CAN.cpp

```diff
@@ -0,0 +1,181 @@
+const AP_Param::GroupInfo AP_BoardConfig_CAN::var_info[] = {
+
+#if HAL_WITH_UAVCAN
+
+#if MAX_NUMBER_OF_CAN_INTERFACES > 0
+    // @Group: P1_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can[0], "P1_", 1, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_INTERFACES > 1
+    // @Group: P2_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can[1], "P2_", 2, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_INTERFACES > 2
+    // @Group: P3_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can[2], "P3_", 3, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_DRIVERS > 0
+    // @Group: D1_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can_protocol[0], "D1_", 4, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_protocol_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_DRIVERS > 1
+    // @Group: D2_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can_protocol[1], "D2_", 5, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_protocol_var_info),
+#endif
+
+#if MAX_NUMBER_OF_CAN_DRIVERS > 2
+    // @Group: D3_
+    // @Path: ../AP_BoardConfig/canbus.cpp
+    AP_SUBGROUPINFO(_var_info_can_protocol[2], "D3_", 6, AP_BoardConfig_CAN, AP_BoardConfig_CAN::CAN_protocol_var_info),
+#endif
+
+#endif
+
+    AP_GROUPEND
+};
+
+#if HAL_WITH_UAVCAN
+int8_t AP_BoardConfig_CAN::_st_can_enable[MAX_NUMBER_OF_CAN_INTERFACES];
+int8_t AP_BoardConfig_CAN::_st_can_debug[MAX_NUMBER_OF_CAN_INTERFACES];
+#endif
+
+void AP_BoardConfig_CAN::init()
+{
+#if HAL_WITH_UAVCAN
+    for (uint8_t i = 0; i < MAX_NUMBER_OF_CAN_INTERFACES; i++)
+    {
+        _st_can_enable[i] = (int8_t) _var_info_can[i]._can_enable;
+        _st_can_debug[i] = (int8_t) _var_info_can[i]._can_debug;
+    }
+#endif
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN
+    px4_setup_canbus();
+#endif
+
+}
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN
+/*
+  setup CANBUS drivers
+ */
+void AP_BoardConfig_CAN::px4_setup_canbus(void)
+{
+#if HAL_WITH_UAVCAN
+    // Create all drivers that we need
+    for (uint8_t i = 0; i < MAX_NUMBER_OF_CAN_INTERFACES; i++)
+    {
+        uint8_t drv_num = _var_info_can[i]._can_enable;
+
+        if (drv_num) {
+            if (hal.can_mgr[drv_num - 1] == nullptr)
+            {
+                const_cast <AP_HAL::HAL&> (hal).can_mgr[drv_num - 1] = new PX4::PX4CANManager;
+            }
+        }
+    }
+
+    bool initret = true;
+
+    for (uint8_t i = 0; i < MAX_NUMBER_OF_CAN_INTERFACES; i++)
+    {
+        uint8_t drv_num = _var_info_can[i]._can_enable;
+
+        if (drv_num) {
+            if (hal.can_mgr[drv_num - 1] != nullptr)
+            {
+                initret &= hal.can_mgr[drv_num - 1]->begin(_var_info_can[i]._can_bitrate, i);
```

Comment:
Ahhh. That's a good question and sure it is not clear enough.
Simple answer: because this loop depends on what is done in previous.
Same driver (manager) can be used for several physical interfaces. So first we need to go through all interfaces and create all managers.
The number of managers can not be more than number of physical interfaces.
Next for the created managers we need to link them to the interfaces.